### PR TITLE
Typing Timer update

### PR DESCRIPTION
**Pull request by singhk97**
Made a few changes to the TypingTimer based on @vule-msft's suggestions

* It implements the `IDisposable` interface to follow C# conventions on object that need to have resources explicitly disposed by the users.
* Implemented a destructor to properly dispose the timer resource when garbage collector destroys this object
* Made timer property private to avoid unexpected mutations from outside this class

### CODE REVIEW

File: dotnet/packages/Microsoft.Bot.Builder.M365/Microsoft.Bot.Builder.M365/TypingTimer.cs [hunk 1]
```diff
@@ -11,9 +11,9 @@ namespace Microsoft.Bot.Builder.M365
     /// <summary>
     /// Encapsulates the logic for sending "typing" activity to the user.
     /// </summary>
-    public class TypingTimer
+    public class TypingTimer : IDisposable
```
[the remark this answers is not in the document]
Comment (singhk97) [reply dotnet/packages/Microsoft.Bot.Builder.M365/Microsoft.Bot.Builder.M365/TypingTimer.cs:8]: The typing indicator experience as a user messaging the bot seems quite natural. The JS SDK sets it as 1s so for parity sakes I'll keep it at this if it's not too big of an issue. What do you say?

File: dotnet/packages/Microsoft.Bot.Builder.M365/Microsoft.Bot.Builder.M365/TypingTimer.cs [hunk 2]
```diff
@@ -38,25 +38,35 @@ public TypingTimer(int interval = 1000)
...
         {
-            if (turnContext.Activity.Type != ActivityTypes.Message || timer != null) return;
+            if (turnContext.Activity.Type != ActivityTypes.Message || _timer != null) return;
 
             // Listen for outgoing activities
             turnContext.OnSendActivities(StopTimerWhenSendMessageActivityHandler);
             
             // Start periodically send "typing" activity
-            timer = new Timer(SendTypingActivity, turnContext, 0, _interval);
+            _timer = new Timer(SendTypingActivity, turnContext, 0, _interval);
```
[the remark this answers is not in the document]
Comment (singhk97) [reply dotnet/packages/Microsoft.Bot.Builder.M365/Microsoft.Bot.Builder.M365/TypingTimer.cs:47]: That's a good point. Looking into it.

[the remark this answers is not in the document]
Comment (singhk97) [reply dotnet/packages/Microsoft.Bot.Builder.M365/Microsoft.Bot.Builder.M365/TypingTimer.cs]: Added functionality + unit test

File: dotnet/packages/Microsoft.Bot.Builder.M365/Microsoft.Bot.Builder.M365/TypingTimer.cs [hunk 3]
```diff
@@ -36,56 +40,98 @@ public TypingTimer(int interval = 1000)
...
+            _timerState = new TimerState(turnContext, false);
 
             // Listen for outgoing activities
             turnContext.OnSendActivities(StopTimerWhenSendMessageActivityHandler);
-            
+
             // Start periodically send "typing" activity
-            timer = new Timer(SendTypingActivity, turnContext, 0, _interval);
-        }
+            _timer = new Timer(SendTypingActivity, _timerState, 0, _interval);
```
Comment (vule-msft) [inline dotnet/packages/Microsoft.Bot.Builder.M365/Microsoft.Bot.Builder.M365/TypingTimer.cs:54]: I believe we can simplify much of this code by using the Change() method inside of the SendTypingActivity callback

```
_timer = new Timer(SendTypingActivity, null, Timeout.Infinite, Timeout.Infinite);
_timer.Change(0,Timeout.Infinite); // fire first time

// inside of callback
private async void SendTypingActivity(object state){
 
  // send and await typing indicator
  // make sure errors are handled if necessary
   await turnContext.SendActivityAsync(new Activity { Type = ActivityTypes.Typing });
  
  // wait 1 sec and fire again
  _timer.Change(1000, Timeout.Infinite);
}

```

Comment (singhk97) [reply dotnet/packages/Microsoft.Bot.Builder.M365/Microsoft.Bot.Builder.M365/TypingTimer.cs]: fixed

**Review by vule-msft — APPROVED** [review]

**Pull request merged**